### PR TITLE
Feature/phenotypefeatureupdates

### PR DESCRIPTION
**Pull request by aparton**

### CODE REVIEW

File: modules/Bio/EnsEMBL/VEP/Pipeline/DumpVEP/Dumper/Variation.pm [hunk 1]
```diff
@@ -388,6 +389,76 @@ sub freqs_from_vcf {
...
+  foreach my $attr(@{$attrib}){
+    $per_allele_hash->{$attr->{risk_allele}}->{$attr->{clinvar_clin_sig}} = 1 if defined($attr->{risk_allele}) ;
+  }
+
+  my @array = keys(%$per_allele_hash);
+  my $output_string = '';
+  foreach my $allele(@array)
+  {
+    my @clinsigarray = keys(%{$per_allele_hash->{$allele}});
+    $output_string .= $allele.':' . (join ',', @clinsigarray) . ';' if scalar(@clinsigarray);
```
Comment (at7) [inline modules/Bio/EnsEMBL/VEP/Pipeline/DumpVEP/Dumper/Variation.pm]: Are we ending up with one too many ';'? What happens next with $output_string?

File: modules/Bio/EnsEMBL/VEP/Pipeline/DumpVEP/Dumper/Variation.pm [hunk 2]
```diff
@@ -388,6 +389,76 @@ sub freqs_from_vcf {
     }
   }
 }
+sub get_pf_features_by_location {
```
Comment (at7) [inline modules/Bio/EnsEMBL/VEP/Pipeline/DumpVEP/Dumper/Variation.pm]: The name is confusing. I would prefer get_phenotype_feature_attribs_by_location

File: modules/Bio/EnsEMBL/VEP/Pipeline/DumpVEP/Dumper/Variation.pm [hunk 3]
```diff
@@ -388,6 +389,76 @@ sub freqs_from_vcf {
...
+
+  # put into a hash so we can lookup by pos
+  my $vep_params = $self->get_vep_params();
+
+  # make sure to include failed variants!
+  $vep_params->{failed} = 1;
+  
+  my $config = Bio::EnsEMBL::VEP::Config->new($vep_params);
+  
+  my $region_size = $self->param('region_size');
```
Comment (at7) [inline modules/Bio/EnsEMBL/VEP/Pipeline/DumpVEP/Dumper/Variation.pm]: where are config and region_size needed again?

File: modules/Bio/EnsEMBL/VEP/Pipeline/DumpVEP/Dumper/Variation.pm [hunk 4]
```diff
@@ -388,6 +389,76 @@ sub freqs_from_vcf {
...
+  my $hive_dbc = $self->dbc;
```
Comment (at7) [inline modules/Bio/EnsEMBL/VEP/Pipeline/DumpVEP/Dumper/Variation.pm]: are you concerned that we are loosing the connection to the hive database? What about connection to the variation database?

File: modules/Bio/EnsEMBL/VEP/Pipeline/DumpVEP/Dumper/Variation.pm [hunk 5]
```diff
@@ -388,6 +389,76 @@ sub freqs_from_vcf {
...
+  
+    my $config = Bio::EnsEMBL::VEP::Config->new($vep_params);  
+    my $as = Bio::EnsEMBL::VEP::AnnotationSource::Database::Variation->new({
+      config => $config,
+      cache_region_size => $region_size,
+    });
+
+    my $pfas = $as->get_adaptor('variation', 'variation')->db->get_PhenotypeFeatureAdaptor()->get_PhenotypeFeatureAttribs_by_location($sr, $region_start, $region_end) if defined($as->get_adaptor('variation', 'variation')->db);
+
+    $self->{pfa_cache}->{$sr . ':' . $region_start . '-' . $region_end} = $pfas;;
```
Comment (at7) [inline modules/Bio/EnsEMBL/VEP/Pipeline/DumpVEP/Dumper/Variation.pm]: one two many `;`

**Comment by sarahhunt** [comment]
I don't think this would work for non-cache users.

Would it work to put a range query on Phenotype features and attribs, then caching of phenotype data on the VF round about here:

https://github.com/Ensembl/ensembl-vep/blob/release/96/modules/Bio/EnsEMBL/VEP/AnnotationSource/Database/Variation.pm#L181

**Comment by aparton** [comment]
It does work for non-cache users, although the code to handle that is currently in the OutputFactory. You're right, it shouldn't be there, I'll do some refactoring.

File: modules/Bio/EnsEMBL/VEP/AnnotationSource/Database/Variation.pm [hunk 6]
```diff
@@ -179,11 +179,31 @@ sub get_features_by_regions_uncached {
...
+    my @vars;
     while($sth->fetch) {
       my %v_copy = %v;
       $v_copy{allele_string} =~ s/\s+/\_/g;
+      my @pf_matched = grep {$_->{_object_id} eq $v_copy{variation_name}} @{$pfs};
+      my @pfas_by_allele;
+      my %clin_sigs;
+      foreach my $pf(@pf_matched)
+      {
+        my $attrib = $pf->get_all_attributes();
```
Comment (sarahhunt) [inline modules/Bio/EnsEMBL/VEP/AnnotationSource/Database/Variation.pm]: Isn't this slow compared to getting all attribs for the slice?

File: modules/Bio/EnsEMBL/VEP/Pipeline/DumpVEP/Dumper/Variation.pm [hunk 7]
```diff
@@ -228,9 +228,10 @@ sub dump_obj {
       defined($v->{minor_allele_freq}) && $v->{minor_allele_freq} =~ /^[0-9\.]+$/ ? sprintf("%.4f", $v->{minor_allele_freq}) : '',
       $v->{clin_sig} || '',
       $v->{phenotype_or_disease} == 0 ? '' : $v->{phenotype_or_disease},
+      $self->get_allele_specific_clin_sigs($sr, $v->{start}, $v->{end}) || '',
```
Comment (sarahhunt) [inline modules/Bio/EnsEMBL/VEP/Pipeline/DumpVEP/Dumper/Variation.pm]: Why is this needed? Aren't these available from the get_features_by_regions_uncached look up?

File: modules/Bio/EnsEMBL/VEP/AnnotationSource/Database/Variation.pm [hunk 8]
```diff
@@ -178,12 +178,28 @@ sub get_features_by_regions_uncached {
     my ($var_id, %vars_by_id);
     $sth->bind_col(1, \$var_id);
     $sth->bind_col($_+2, \$v{$VAR_CACHE_COLS[$_]}) for (0..$#VAR_CACHE_COLS);
+    
+    my $adaptor = $self->get_adaptor('variation', 'phenotypefeature');
+    my $pfs = $adaptor->get_PhenotypeFeatures_by_location($chr_is_seq_region ? $chr : $sr_cache->{$chr}, $s, $e, 32) if defined($adaptor); # 32 is the sourceID for Clinvar PFs
```
Comment (sarahhunt) [inline modules/Bio/EnsEMBL/VEP/AnnotationSource/Database/Variation.pm]: This is a slow query and the results aren't being used.

File: modules/Bio/EnsEMBL/VEP/AnnotationSource/Database/Variation.pm [hunk 9]
```diff
@@ -178,12 +178,28 @@ sub get_features_by_regions_uncached {
...
+    my $attribs = $adaptor->get_PhenotypeFeatureAttribs_by_location($sr_cache->{$chr}, $s, $e) if defined($adaptor);
```
Comment (sarahhunt) [inline modules/Bio/EnsEMBL/VEP/AnnotationSource/Database/Variation.pm]: Restricting this one to ClinVar only should provide a speed up

File: modules/Bio/EnsEMBL/VEP/AnnotationSource/Database/Variation.pm (hunk 8 above)
Comment (ens-lgil) [inline modules/Bio/EnsEMBL/VEP/AnnotationSource/Database/Variation.pm]: I am not a big fan of hard coded source IDs.
They are quite stable but they are not a stable ID.
Furthermore, in the future an other species can have phenotype entries on a source with the source ID 32.

File: modules/Bio/EnsEMBL/VEP/AnnotationSource/Database/Variation.pm [hunk 10]
```diff
@@ -180,7 +180,8 @@ sub get_features_by_regions_uncached {
     $sth->bind_col($_+2, \$v{$VAR_CACHE_COLS[$_]}) for (0..$#VAR_CACHE_COLS);
     
     my $adaptor = $self->get_adaptor('variation', 'phenotypefeature');
-    my $attribs = $adaptor->get_clinsig_alleles_by_location($chr_is_seq_region ? $chr : $sr_cache->{$chr}, $s, $e) if defined($adaptor);
+    my $source_id = $self->clinvar_source_id_cache;
+    my $attribs = $adaptor->get_clinsig_alleles_by_location($chr_is_seq_region ? $chr : $sr_cache->{$chr}, $s, $e, $source_id) if defined($adaptor);
```
Comment (ens-lgil) [inline modules/Bio/EnsEMBL/VEP/AnnotationSource/Database/Variation.pm]: Should we add a check that `$source_id` is defined as well ?
e.g. `...  if ( defined($adaptor) && defined($source_id) );`

Comment (aparton) [reply modules/Bio/EnsEMBL/VEP/AnnotationSource/Database/Variation.pm]: Yes. Done.
